src/main/scala/fr/acinq/bitcoin/Psbt.scala

```diff
@@ -0,0 +1,377 @@
+
+    private def readOutputs(input: InputStream, expectedCount: Int): Try[Seq[PartiallySignedOutput]] = trySequence((0 until expectedCount).map(_ => readOutput(input)))
+
+    private def readOutput(input: InputStream): Try[PartiallySignedOutput] = readDataMap(input).flatMap(entries => {
+      val keyTypes = Set(0x00, 0x01, 0x02).map(_.toByte)
+      val (known, unknown) = entries.partition(entry => entry.key.headOption.exists(keyTypes.contains))
+      val redeemScript_opt: Try[Option[List[ScriptElt]]] = known.find(_.key.head == 0x00).map {
+        case DataEntry(key, _) if key.length != 1 => Failure(new IllegalArgumentException("psbt redeem script key must contain exactly 1 byte"))
+        case DataEntry(_, value) => Success(Some(Script.parse(value)))
+      }.getOrElse(Success(None))
+      val witnessScript_opt: Try[Option[List[ScriptElt]]] = known.find(_.key.head == 0x01).map {
+        case DataEntry(key, _) if key.length != 1 => Failure(new IllegalArgumentException("psbt witness script key must contain exactly 1 byte"))
+        case DataEntry(_, value) => Success(Some(Script.parse(value)))
+      }.getOrElse(Success(None))
+      val derivationPaths_opt: Try[Map[PublicKey, KeyPathWithMaster]] = trySequence(known.filter(_.key.head == 0x02).map {
+        case DataEntry(_, value) if value.length < 4 || value.length % 4 != 0 => Failure(new IllegalArgumentException("psbt bip32 derivation must contain master key fingerprint and child indexes"))
+        case DataEntry(key, value) =>
+          val publicKey = PublicKey(ByteVector(key.tail), checkValid = true)
+          val masterKeyFingerprint = Protocol.uint32(value.take(4), ByteOrder.BIG_ENDIAN)
+          val childCount = (value.length / 4) - 1
+          val derivationPath = KeyPath((0 until childCount).map(i => Protocol.uint32(value.slice(4 * (i + 1), 4 * (i + 2)), ByteOrder.LITTLE_ENDIAN)))
+          Success(publicKey, KeyPathWithMaster(masterKeyFingerprint, derivationPath))
+      }).map(_.toMap)
+      for {
+        redeemScript <- redeemScript_opt
+        witnessScript <- witnessScript_opt
+        derivationPaths <- derivationPaths_opt
+      } yield PartiallySignedOutput(redeemScript, witnessScript, derivationPaths, unknown)
+    })
+
+    @tailrec
+    private def readDataMap(input: InputStream, entries: Seq[DataEntry] = Nil): Try[Seq[DataEntry]] = readDataEntry(input) match {
+      case Success(Some(entry)) => readDataMap(input, entry +: entries)
+      // NB: we convert Array[Byte] to ByteVector because comparing two Array[Byte] doesn't compare the contents of the arrays.
+      case Success(None) => if (entries.map(entry => ByteVector(entry.key)).toSet.size != entries.size) {
+        Failure(new IllegalArgumentException("psbt must not contain duplicate keys"))
+      } else {
+        Success(entries)
+      }
+      case Failure(ex) => Failure(ex)
+    }
+
+    private def readDataEntry(input: InputStream): Try[Option[DataEntry]] = Protocol.varint(input) match {
```

Comment:
Good catch, I need to clean that up.

Reply:
Done in https://github.com/ACINQ/bitcoin-lib/pull/42/commits/bbe9fd6b3c8c1ff91f98126b34545ea1b6af3b30
Note that I don't need to wrap every `readXXX` function in a `Try` because they're using a `flatMap` on a `Try`, so they automatically wrap uncaught exceptions in a `Failure`.